File: src/f2a/core/schema.py
"""데이터 스키마 추론 및 관리."""

from __future__ import annotations

from dataclasses import dataclass, field

import pandas as pd

from f2a.utils.type_inference import ColumnType, infer_all_types


@dataclass
class ColumnInfo:
    """개별 컬럼에 대한 메타 정보."""

    name: str
    dtype: str
    inferred_type: ColumnType
    n_unique: int
    n_missing: int
    missing_ratio: float


@dataclass
class DataSchema:
    """데이터프레임 전체의 스키마 정보."""

    n_rows: int
    n_cols: int
    columns: list[ColumnInfo] = field(default_factory=list)
    memory_usage_mb: float = 0.0

    @property
    def numeric_columns(self) -> list[str]:
        """수치형 컬럼 이름 목록."""
        return [c.name for c in self.columns if c.inferred_type == ColumnType.NUMERIC]

    @property
    def categorical_columns(self) -> list[str]:
        """범주형 컬럼 이름 목록."""
        return [c.name for c in self.columns if c.inferred_type == ColumnType.CATEGORICAL]

    @property
    def text_columns(self) -> list[str]:
        """텍스트 컬럼 이름 목록."""
        return [c.name for c in self.columns if c.inferred_type == ColumnType.TEXT]

    @property
    def datetime_columns(self) -> list[str]:
        """일시 컬럼 이름 목록."""
        return [c.name for c in self.columns if c.inferred_type == ColumnType.DATETIME]

    def summary_dict(self) -> dict[str, str | int | float]:
        """스키마 요약을 딕셔너리로 반환합니다."""
        return {
            "rows": self.n_rows,
            "columns": self.n_cols,
            "numeric": len(self.numeric_columns),
            "categorical": len(self.categorical_columns),
            "text": len(self.text_columns),
            "datetime": len(self.datetime_columns),
            "memory_mb": round(self.memory_usage_mb, 2),
        }


def infer_schema(df: pd.DataFrame) -> DataSchema:
    """DataFrame으로부터 스키마를 추론합니다.

    Args:
        df: 분석 대상 DataFrame.

    Returns:
        추론된 :class:`DataSchema`.
    """
    type_map = infer_all_types(df)
    columns: list[ColumnInfo] = []

    for col in df.columns:
        n_missing = int(df[col].isna().sum())
        columns.append(
            ColumnInfo(
                name=col,
                dtype=str(df[col].dtype),
                inferred_type=type_map[col],
                n_unique=int(df[col].nunique()),
                n_missing=n_missing,
                missing_ratio=round(n_missing / len(df), 4) if len(df) > 0 else 0.0,
            )
        )

    memory_mb = df.memory_usage(deep=True).sum() / (1024 * 1024)

    return DataSchema(
        n_rows=len(df),
        n_cols=len(df.columns),
        columns=columns,
        memory_usage_mb=round(memory_mb, 2),
    )

File: src/core/schema.rs
use polars::prelude::*;
use serde::{Deserialize, Serialize};

use crate::utils::types::{infer_column_type, ColumnType};

// ─── Per-column metadata ────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnInfo {
    pub name: String,
    pub dtype: String,
    pub inferred_type: ColumnType,
    pub n_unique: usize,
    pub n_missing: usize,
    pub missing_ratio: f64,
}

// ─── Dataset-level schema ───────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSchema {
    pub n_rows: usize,
    pub n_cols: usize,
    pub columns: Vec<ColumnInfo>,
    pub memory_usage_bytes: usize,
}

impl DataSchema {
    /// Build a `DataSchema` from a Polars `DataFrame`.
    pub fn from_dataframe(df: &DataFrame) -> Self {
        let n_rows = df.height();
        let n_cols = df.width();

        let columns: Vec<ColumnInfo> = df
            .get_columns()
            .iter()
            .map(|col| {
                let name = col.name().to_string();
                let dtype = format!("{:?}", col.dtype());
                let n_missing = col.null_count();
                let missing_ratio = if n_rows > 0 {
                    n_missing as f64 / n_rows as f64
                } else {
                    0.0
                };

                let n_unique = col.n_unique().unwrap_or(0);

                // Compute avg string length for text classification heuristic
                let avg_str_len = if col.dtype() == &DataType::String {
                    let lengths = col.str().ok().map(|ca| {
                        let total: usize =
                            ca.into_iter().filter_map(|opt| opt.map(|s| s.len())).sum();
                        let count = ca.into_iter().filter(|o| o.is_some()).count();
                        if count > 0 {
                            total as f64 / count as f64
                        } else {
                            0.0
                        }
                    });
                    lengths
                } else {
                    None
                };

                let inferred_type = infer_column_type(col.dtype(), n_unique, n_rows, avg_str_len);

                ColumnInfo {
                    name,
                    dtype,
                    inferred_type,
                    n_unique,
                    n_missing,
                    missing_ratio,
                }
            })
            .collect();

        // Rough memory estimation
        let memory_usage_bytes = df.estimated_size();

        DataSchema {
            n_rows,
            n_cols,
            columns,
            memory_usage_bytes,
        }
    }

    // ── Convenience accessors ───────────────────────────────────

    pub fn numeric_columns(&self) -> Vec<&str> {
        self.columns
            .iter()
            .filter(|c| c.inferred_type == ColumnType::Numeric)
            .map(|c| c.name.as_str())
            .collect()
    }

    pub fn categorical_columns(&self) -> Vec<&str> {
        self.columns
            .iter()
            .filter(|c| c.inferred_type == ColumnType::Categorical)
            .map(|c| c.name.as_str())
            .collect()
    }

    pub fn text_columns(&self) -> Vec<&str> {
        self.columns
            .iter()
            .filter(|c| c.inferred_type == ColumnType::Text)
            .map(|c| c.name.as_str())
            .collect()
    }

    pub fn datetime_columns(&self) -> Vec<&str> {
        self.columns
            .iter()
            .filter(|c| c.inferred_type == ColumnType::DateTime)
            .map(|c| c.name.as_str())
            .collect()
    }

    pub fn boolean_columns(&self) -> Vec<&str> {
        self.columns
            .iter()
            .filter(|c| c.inferred_type == ColumnType::Boolean)
            .map(|c| c.name.as_str())
            .collect()
    }

    pub fn column_info(&self, name: &str) -> Option<&ColumnInfo> {
        self.columns.iter().find(|c| c.name == name)
    }
}
